contracts/eos/verifier/src/cpp/verifier.cpp

```diff
@@ -0,0 +1,170 @@
+    data_index verifier_data(get_self(), get_first_receiver().value);
+    auto iterator = verifier_data.find(verifier_name.value);
+
+    check(iterator != verifier_data.end(),
+          "DendrETH verifier not instantiated");
+    // Prepare data for the nim verifier function
+    std::array<uint8_t, 1152> _vk;
+    std::array<uint8_t, 384> _prf;
+    std::array<uint8_t, 32> _current_header_root;
+    std::array<uint8_t, 32> _new_optimistic_header_root;
+    std::array<uint8_t, 32> _new_finalized_header_root;
+    std::array<uint8_t, 32> _new_execution_state_root;
+
+    verifier_data.modify(iterator, verifier_name, [&](auto &row) {
+      std::copy(row.vk.begin(), row.vk.end(), _vk.begin());
+      std::copy(proof.begin(), proof.end(), _prf.begin());
+      std::copy(row.new_optimistic_header_roots[row.current_index].begin(),
+                row.new_optimistic_header_roots[row.current_index].end(),
+                _current_header_root.begin());
+      std::copy(new_optimistic_header_root.begin(),
+                new_optimistic_header_root.end(),
+                _new_optimistic_header_root.begin());
+      std::copy(new_finalized_header_root.begin(),
+                new_finalized_header_root.end(),
+                _new_finalized_header_root.begin());
+      std::copy(new_execution_state_root.begin(),
+                new_execution_state_root.end(),
+                _new_execution_state_root.begin());
+
+      bool result = makePairsAndVerify(
+          &_vk, &_prf, &_current_header_root, &_new_optimistic_header_root,
+          &_new_finalized_header_root, &_new_execution_state_root);
+
+      check(result, "Verification failed. Incorrect update");
+      if (row.current_index == 31) {
+        row.current_index = 0;
+      } else {
+        row.current_index++;
+      }
+
+      std::vector<uint8_t> _new_current_header_root(
+          _current_header_root.begin(), _current_header_root.end());
+
+      row.new_optimistic_header_roots[row.current_index] =
+          _new_current_header_root;
+      row.new_finalized_header_roots[row.current_index] =
+          new_finalized_header_root;
+      row.new_execution_state_roots[row.current_index] =
+          new_execution_state_root;
+    });
+  }
+
+  [[eosio::action]] void printheader() {
+    data_index verifier_data(get_self(), get_first_receiver().value);
+    auto &result = verifier_data.get(verifier_name.value);
+    verifier_data.modify(result, verifier_name, [&](auto &row) {
+      std::array<uint8_t, 32> _current_header_root;
+
+      std::copy(row.new_optimistic_header_roots[row.current_index].begin(),
+                row.new_optimistic_header_roots[row.current_index].end(),
+                _current_header_root.begin());
+      eosio::print("[");
+      for (int i = 0; i < 32; i++) {
+        eosio::print(_current_header_root[i]);
+        if (i != 31) {
+          eosio::print(",");
+        }
+      }
+      eosio::print("]");
```

Comment:
This can be extracted as separate function that you can call here and below in `printraders`

Reply:
Doing that is probably not a good idea as we would need to pass 'row' to that function and its type makes this hard and likely expensive.

Reply:
Disagree. You only need the array with the copy of what is in the row. Smt like this:
```
void printArray(srd::array<uint8_t, 32> array) {
      eosio::print("[");
      for (int i = 0; i < 32; i++) {
        eosio::print(array[i]);
        if (i != 31) {
          eosio::print(",");
        }
      }
      eosio::print("]");
```